pkg/store/proxy.go

```diff
@@ -50,13 +50,127 @@ type Client interface {
 type ProxyStore struct {
 	logger         log.Logger
 	stores         func() []Client
+	rules          func() []storepb.RulesClient
 	component      component.StoreAPI
 	selectorLabels labels.Labels
 
 	responseTimeout time.Duration
 	metrics         *proxyStoreMetrics
 }
 
+func (s *ProxyStore) Rules(req *storepb.RulesRequest, srv storepb.Rules_RulesServer) error {
+	var (
+		g, gctx  = errgroup.WithContext(srv.Context())
+		respChan = make(chan *storepb.RuleGroup, 10)
+		groups   []*storepb.RuleGroup
+	)
+
+	defer func() { close(respChan) }()
+
+	for _, rulesClient := range s.rules() {
+		rs := &rulesStream{
+			client:  rulesClient,
+			request: req,
+			channel: respChan,
+			server:  srv,
+		}
+		g.Go(func() error { return rs.receive(gctx) })
+	}
+
+	for resp := range respChan {
```

Comment:
Does it work? What happens when all of the senders finish sending their data? AFAICT no one closes this channel. My tests also show that goroutines get stuck on:

```
16 @ 0x437580 0x405f17 0x405c1b 0x16df6f0 0xafbe49 0x1891f6e 0xe2a82f 0xe306c4 0xe32474 0xe2a82f 0x188485b 0xe2a82f 0xe2a9ef 0xab9dee 0xabbf76 0xac9de1 0x4675e1
#	0x16df6ef	github.com/thanos-io/thanos/pkg/store.(*ProxyStore).Rules+0x2cf						/home/gstatkevicius/dev/thanos/pkg/store/proxy.go:80
#	0xafbe48	github.com/thanos-io/thanos/pkg/store/storepb._Rules_Rules_Handler+0x108				/home/gstatkevicius/dev/thanos/pkg/store/storepb/rpc.pb.go:1210
#	0x1891f6d	github.com/grpc-ecosystem/go-grpc-middleware/recovery.StreamServerInterceptor.func1+0xad		/home/gstatkevicius/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/recovery
```

Reply:
yes, i just found it too :+1: added initial e2e tests to cover this. we need to start a goroutine and wait for the errgroup to finish before closing the response channel.

Reply:
rationale of why you observed the above: effectively the response channel never got closed ;-)